src/manifest/owned/pfm.rs

```diff
@@ -102,40 +93,205 @@ pub enum Element {
+    pub updates_take_effect: UpdatesTakeEffect,
+    pub raw_flags: u8,
+}
+
+impl FirmwareFlags {
+    fn new(raw_flags: u8) -> Self {
+        Self {
+            updates_take_effect: UpdatesTakeEffect::new(raw_flags),
+            raw_flags,
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+impl Serialize for FirmwareFlags {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        crate::serde::se_bin(self.raw_flags, serializer)
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> Deserialize<'de> for FirmwareFlags {
+    fn deserialize<D>(deserializer: D) -> Result<FirmwareFlags, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let raw_flags = crate::serde::de_radix(deserializer)?;
+        Ok(FirmwareFlags::new(raw_flags))
+    }
+}
+
+
+/// Updates Take Effect
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub enum UpdatesTakeEffect {
+    /// Only on reboot
+    OnlyOnReboot,
+    /// At run-time and on reboot
+    AtRuntimeAndOnReboot,
+}
+
+impl UpdatesTakeEffect {
+    fn new(raw_flags: u8) -> Self {
+        if zeroth_bit_zero(raw_flags) {
+            Self::OnlyOnReboot
+        } else {
+            Self::AtRuntimeAndOnReboot
+        }
+    }
+}
+
 /// A read-write region.
 #[allow(missing_docs)]
 #[derive(Clone, PartialEq, Eq, Debug, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct Rw {
-    #[cfg_attr(
-        feature = "serde",
-        serde(
-            deserialize_with = "crate::serde::de_radix",
-            serialize_with = "crate::serde::se_bin",
-        )
-    )]
-    pub flags: u8,
+    pub flags: RwFlags,
     pub region: Region,
 }
 
+/// R/W Flags
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct RwFlags {
+    /// Operation on Failure
+    pub operation_on_failure: OperationOnFailure,
+    /// The raw flag value
+    pub raw_flags: u8,
+}
+
+impl RwFlags {
+    fn new(raw_flags: u8) -> Self {
+        Self {
+            operation_on_failure: OperationOnFailure::new(raw_flags),
+            raw_flags,
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+impl Serialize for RwFlags {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        crate::serde::se_bin(self.raw_flags, serializer)
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> Deserialize<'de> for RwFlags {
+    fn deserialize<D>(deserializer: D) -> Result<RwFlags, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let raw_flags = crate::serde::de_radix(deserializer)?;
+        Ok(RwFlags::new(raw_flags))
+    }
+}
+
+/// Operation on Failure
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub enum OperationOnFailure {
+    /// Do nothing
+    DoNothing,
+    /// Restore region from RO
+    RestoreRegionFromRo,
+    /// Erase region
+    EraseRegion,
+}
+
+impl OperationOnFailure {
+    fn new(raw_flags: u8) -> Self {
+        match raw_flags << 6 >> 6 {
```

Comment:
I'd have written `raw_flags & !0b11`. The shifts aren't super easy to read imo.

Reply:
resolving, outdated